capsules/src/analog_comparator.rs

```diff
@@ -117,7 +124,24 @@ impl<'a, A: hil::analog_comparator::AnalogComparator<'a>> Driver for AnalogCompa
     /// - `3`: Stop interrupt-based comparisons.
     ///        Input x chooses the desired comparator ACx (e.g. 0 or 1 for
     ///        hail, 0-3 for imix)
-    fn command(&self, command_num: usize, channel: usize, _: usize, _: AppId) -> CommandReturn {
+    fn command(&self, command_num: usize, channel: usize, _: usize, appid: AppId) -> CommandReturn {
+        if command_num == 0 {
+            // Handle this first as it should be returned unconditionally.
+            return CommandReturn::success_u32(self.channels.len() as u32);
+        }
+
+        // Check if this driver is free, or already dedicated to this process.
+        let match_or_empty_or_nonexistant = self.current_process.map_or(true, |current_process| {
+            self.grants
+                .enter(*current_process, |_, _| current_process == &appid)
+                .unwrap_or(true)
+        });
+        if match_or_empty_or_nonexistant {
+            self.current_process.set(appid);
+        } else {
+            return CommandReturn::failure(ErrorCode::NOMEM);
```

Comment:
Intuitively I would've expected another ErrorCode, something like `BUSY`? `NOMEM` kind of fits here I guess?

Reply:
Historically that is the error that has been returned because there was no memory available to store the new process's request when grants were not being used.

Maybe we should use `ErrorCode::RESERVE`? I'm not that worried either way as I don't think userspace will actually use these and they should be virtualized anyway.

Reply:
The explanation behind `NOMEM` makes sense, I'm happy to use that. I went through the other error codes and I agree, no other code seems to fit well.

I'll update #2518 for consistency's sake.